Created stubs for the OCI8 extension.

jensdenies opened the pull request:
Hi,

I noticed the OCI8 extension didn't have stubs yet, that's why i decided to try to add them. I went over the implementation of the functions and methods to determine the types, but it would be nice if someone could double check them.

I've ran the tests and about 78% passed, the ones that failed didn't seem to be related to my changes. I could be wrong, however, so it would be nice if someone could check whether the functions and classes still work as expected.

If i recall correctly, someone mentioned the addition of the "false" return type. Did the "false" return type make it into PHP8? If so, should i add return types for the functions which return false and something else?

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+ * @param int $session_mode
+ * @return resource|false @todo double check
+ */
+function oci_new_connect(string $username, string $password, ?string $connection_string = null, ?string $character_set = null, int $session_mode = OCI_DEFAULT) {}
```
`$connection_string` and `$character_set` are not nullable

jensdenies replied:
@kocsismate Do you happen to know the default value for these parameters?

kocsismate replied:
I believe both are `UNKNOWN`. :)

Off-topic: it's weird for me that the `$connection_string` parameter name refers to the `dbname` C variable. I see that it's also a preexisting stuff, but the parameter name is a bit deceptive.

jensdenies replied:
I removed nullability and made the default value `UNKNOWN`. Thanks!

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+ * @param string|null $connection_string
+ * @param string|null $character_set
+ * @param int $session_mode
+ * @return resource|false @todo double check
```
Yes, this is correct :)

jensdenies replied:
Thanks.

kocsismate commented:
@jensdenies Yes, `false` is a valid union type (together with at least one more type). For more information, have a look at https://wiki.php.net/rfc/union_types_v2

I believe not many people can run these tests so it would be appreciated if you could share some of the failing ones. These are not even run in CI.

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+ * @param int $length
+ * @return int|false @todo return false
+ */
+function oci_lob_write(OCI_Lob $lob_descriptor, string $string, int $length = 0) {}
```
It seems to me that the default value of `$length` is not right here. Even though it's apparently `0`, we can't document this yet, since the behaviour changes depending on whether we provide the 3rd parameter due to the `if (ZEND_NUM_ARGS() == ...` checks). Ideally, the behaviour should depend on the value solely.

I think there's two possibility now: either we should mark the default value as `UNKNOWN` in the stub for now, or we should fix the default value handling. I'm not sure if we can use `0` as a default value because that seems to be a breaking change for me. @nikic please correct me if I'm wrong. Should we use `null` as default? Maybe `-1`?

nikic replied:
I'd go for null default here (or UNKNOWN for now). I looked at what fwrite() does, but unfortunately that one is also arg count dependent, so not a good reference :)

jensdenies replied:
@kocsismate @nikic I made $length nullable and made null the default value, is that ok?

kocsismate replied:
Currently, the default value is neither `0`, nor `null`, not anything else because default value handling is not right in the code :) In these situations we document `UNKNOWN`. In order to be eligible for documenting `null`, the implementation also has to be adapted. Here's an example PR which does the same: https://github.com/php/php-src/pull/5516/

This PR is the part of a crusade which is currently going on against `UNKNOWN` default values. Right now, we have around 280 of such default values, but it used to be even more, probably 400-ish.

So I think it's perfectly ok if you just use `UNKNOWN` for now. However, if you are up for the challenge, you can also fix these after the ZPP-related PR is merged.

jensdenies replied:
I removed nullability and made the default value `UNKNOWN`.

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+<?php
+
```
Thanks, added the annotation and replaced the function entries.

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+
+/**
+ * @param resource $statement_resource
+ * @param string $column_name
```
Done, thanks.

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+
+/**
+ * @param resource $statement_resource
+ * @param mixed $column_number_or_name
```
I've added the mixed type hints, thanks.

jensdenies commented on ext/oci8/oci8.c:
```diff
@@ -847,40 +342,40 @@ static const zend_function_entry php_oci_functions[] = {
-	PHP_FALIAS(save,		oci_lob_save,			arginfo_oci_lob_save_method)
-	PHP_FALIAS(savefile,	oci_lob_import,			arginfo_oci_lob_import_method)
-	PHP_FALIAS(free,		oci_free_descriptor,	arginfo_oci_free_descriptor_method)
+	PHP_FALIAS(load,		oci_lob_load,			arginfo_class_OCI_Lob_load)
```
Done, thanks.

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @return string|false @todo return false
```
Done, thanks.

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @param $filename @todo path?
```
Done, thanks.

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+/**
+ * @param OCI_Lob $lob_descriptor
+ * @param int $length
+ * @return string|false @todo return false
```
Done, thanks.

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+ * @param int $whence
+ * @return bool
+ */
+function oci_lob_seek(OCI_Lob $lob_descriptor, int $offset, int $whence = 0): bool {}
```
Done, thanks.

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+ * @param resource $connection_resource
+ * @return bool
+ */
+function oci_close($connection_resource): bool {}
```
If i place it after ZPP, what should i return instead?

jensdenies replied:
Thanks @kocsismate, this really helps with gaining some more knowledge of the PHP internals. I implemented this in #5714

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+ * @param OCI_Lob $lob_descriptor
+ * @return bool
+ */
+function oci_free_descriptor(OCI_Lob $lob_descriptor): bool {}
```
Could you maybe show me an example? I get what you are saying but i'm not sure how you'd like to see it fixed.

jensdenies replied:
@kocsismate Maybe it's better to fix this in a separate PR?

kocsismate replied:
I added a sample commit for these kind of fixes: https://github.com/php/php-src/commit/793823655e91022f7053ce5d64c7c623e5f9a5c2

But yes, these can be implemented in a separate PR.

jensdenies replied:
Thanks, i implemented this in #5714

jensdenies commented:
@kocsismate Thanks for the kind words and your review. I've processed some of your feedback and replied to some of your comments.

Some failing tests include:

ext/oci8/tests/old_oci_close1.phpt (fatal TypeError instead of expected warning).
ext/oci8/tests/define1.phpt (fatal TypeError instead of expected warning).
ext/oci8/tests/bug42496_1.phpt ( ** ERROR: process timed out **)
ext/oci8/tests/array_bind_float.phpt (expected: 2.5658, actual: 2.5658000000000003)

These are just a few.

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+ * @param int $offset
+ * @return bool
+ */
+function oci_lob_save(OCI_Lob $lob_descriptor, string $data, int $offset = 0): bool {}
```
Thanks, i implemented this in #5714

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+ * @param OCI_Lob $lob_descriptor
+ * @return string|false @todo return false
+ */
+function oci_lob_load(OCI_Lob $lob_descriptor) {}
```
Thanks, i implemented this in #5714

jensdenies commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,734 @@
+ * @param OCI_Lob $lob_descriptor
+ * @return bool
+ */
+function oci_lob_eof(OCI_Lob $lob_descriptor): bool {}
```
Thanks, i implemented this in #5714

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,826 @@
+ * @alias oci_field_name
+ * @deprecated
+ */
+function ocicolumnname($statement_resource, mixed $column_number_or_name) {}
```
same as above

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,826 @@
+ * @param resource $statement_resource
+ * @return int|false|null @todo return false, null?
+ */
+function oci_field_size($statement_resource, mixed $column_number_or_name) {}
```
here (and a few below) as well

kocsismate replied:
also, the return type doesn't include `null`

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,826 @@
+ */
+function ocicolumnisnull($statement_resource, mixed $column_number_or_name): bool {}
+
+function oci_internal_debug(bool $mode): void {}
```
I can't see any parameters. (?)

kocsismate replied:
Shouldn't this be deprecated? It doesn't seem to be very useful anymore.

jensdenies replied:
I've spoken to @cjbj about this and i will make a PR to remove this function after this one has been merged.

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,826 @@
+
+/**
+ * @param resource $statement_resource
+ * @param mixed $result
```
```suggestion
 * @param array $result
```

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,826 @@
+ * @param resource $statement_resource
+ * @param mixed $result
+ */
+function ocifetchinto($statement_resource, &$result, int $mode = 0): int|false {}
```
I'm not 100% percent sure here, but the default value seems to be `OCI_NUM`.

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,826 @@
+
+/**
+ * @param resource $statement_resource
+ * @param mixed $output
```
```suggestion
 * @param array $output
```

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,826 @@
+/**
+ * @param resource $statement_resource
+ */
+function oci_fetch_object($statement_resource, int $mode = 0): object {}
```
```suggestion
function oci_fetch_object($statement_resource, int $mode = PHP_OCI_ASSOC | PHP_OCI_RETURN_NULLS): ?stdClass {}
```

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,826 @@
+/**
+ * @param resource $statement_resource
+ */
+function oci_fetch_array($statement_resource, int $mode = 0): array|false {}
```
```suggestion
function oci_fetch_array($statement_resource, int $mode = PHP_OCI_BOTH | PHP_OCI_RETURN_NULLS): array|false {}
```

jensdenies commented:
@kocsismate I've fixed the things you requested. I also updated the oci8_statement.c file to use `Z_PARAM_STR_OR_LONG`, could you please have a look to check whether it is okay?

I have converted this to a draft PR because i did not run the tests with the newest changes yet.

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+/**
+ * @param resource $connection_resource
+ */
+function oci_set_action($connection_resource, string $action): bool {}
```
Same again. :S

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+/**
+ * @param resource $connection_resource
+ */
+function oci_set_client_info($connection_resource, string $client_information): bool {}
```
And again

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+/**
+ * @param resource $connection_resource
+ */
+function oci_set_db_operation($connection_resource, string $action): bool {}
```
Again...

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+/**
+ * @param resource $connection_resource
+ */
+function oci_set_call_timeout($connection_resource, int $call_timeout): bool {}
```
And again

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+
+/**
+ * @param resource $connection_resource
+ * @return resource|false @todo check return false
```
```suggestion
 * @return resource|false
```

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+
+/**
+ * @param resource $connection_resource
+ * @return resource|false @todo check return false
```
```suggestion
 * @return resource|false
```

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+/**
+ * @param resource $statement_resource
+ */
+function oci_result($statement_resource, mixed $column_number_or_name): mixed {}
```
ZPP for `$column_number_or_name` could be refactored to be `int|string`

kocsismate replied:
The previous `Z_PARAM_STR_OR_LONG` looked good!

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+ */
+function ociresult($statement_resource, mixed $column_number_or_name): mixed {}
+
+function oci_client_version(): string {}
```
ZPP is missing

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+ */
+function ocicollappend(OCI_Collection $collection, string $value): bool {}
+
+function oci_collection_element_get(OCI_Collection $collection, int $index): mixed {}
```
```suggestion
function oci_collection_element_get(OCI_Collection $collection, int $index):string|double|null {}
```
It seems that only these three return types are supported (`php_oci_collection_element_get()` says so) :)

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+ * @alias oci_collection_element_get
+ * @deprecated
+ */
+function ocicollgetelem(OCI_Collection $collection, int $index): mixed {}
```
same as before

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+/**
+ * @param resource $connection_resource
+ */
+function oci_new_collection($connection_resource, string $type_name, ?string $schema_name = null): OCI_Collection|false {}
```
```suggestion
function oci_new_collection($connection_resource, string $type_name, string $schema_name = UNKNOWN): OCI_Collection|false {}
```

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+ * @alias oci_new_collection
+ * @deprecated
+ */
+function ocinewcollection($connection_resource, string $type_name, ?string $schema_name = null): OCI_Collection|false {}
```
same as before

kocsismate commented on ext/oci8/oci8_statement.c:
```diff
@@ -1516,13 +1516,16 @@ sb4 php_oci_bind_out_callback(
-	zval *z_statement, *column_index;
+	zval *z_statement;
+	zend_string *column_index_string = NULL;
+	zend_long column_index_int = 0;
```
No need for the default value

jensdenies commented:
Thanks @kocsismate, i committed the fixes.

kocsismate commented on ext/oci8/oci8_interface.c:
```diff
@@ -2189,6 +2197,7 @@ PHP_FUNCTION(oci_result)
    Return a string containing runtime client library version information */
 PHP_FUNCTION(oci_client_version)
 {
+	ZEND_PARSE_PARAMETERS_NONE();
```
Nit: This would be nicer to be placed after the declaration.

kocsismate replied:
I think it's still not addressed.

jensdenies replied:
Hm, that's strange. If you check my last commit you can see the change.

kocsismate replied:
Hmm, no idea why GitHub shown me the previous state, but I can definitely see it now. :) Thanks!

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+/**
+ * @param resource $connection_resource
+ */
+function oci_register_taf_callback($connection_resource, mixed $function_name): bool {}
```
I'd leave the `mixed` type, and add a `callable` type annotation in PHPDoc

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+     * @alias oci_lob_seek
+     * @return bool
+     */
+    public function seek(int $offset, int $whence = 0) {}
```
```suggestion
    public function seek(int $offset, int $whence = OCI_SEEK_SET) {}
```

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+     * @alias oci_lob_write
+     * @return int|false
+     */
+    public function write(string $string, int $length = 0) {}
```
```suggestion
    public function write(string $string, int $length = UNKNOWN) {}
```
Just like for `oci_lob_write()`

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+     * @alias oci_lob_erase
+     * @return int|false
+     */
+    public function erase(int $offset = -1, int $length = -1) {}
```
Unknown defaults

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+     * @alias oci_lob_export
+     * @return bool
+     */
+    public function writetofile(string $path, int $start = -1, int $length = -1) {}
```
Unknown defaults

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+     * @alias oci_lob_export
+     * @return bool
+     */
+    public function export(string $path, int $start = -1, int $length = -1) {}
```
same here

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,814 @@
+
+    /**
+     * @alias oci_collection_element_get
+     * @return mixed
```
```suggestion
     * @return string|double|null
```

jensdenies commented:
Thanks again @kocsismate. I committed the last fixes @nikic.

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,815 @@
+
+function oci_lob_is_equal(OCI_Lob $lob_descriptor_first, OCI_Lob $lob_descriptor_second): bool {}
+
+function oci_lob_export(OCI_Lob $lob_descriptor, $path, int $start = UNKNOWN, int $length = UNKNOWN): bool {}
```
the `string` type of the `$path` parameter is missing

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,815 @@
+ * @alias oci_lob_export
+ * @deprecated
+ */
+function ociwritelobtofile(OCI_Lob $lob_descriptor, $path, int $start = UNKNOWN, int $length = UNKNOWN): bool {}
```
same as before

jensdenies commented:
@kocsismate Fixed, thanks.

kocsismate commented on ext/oci8/oci8.stub.php:
```diff
@@ -0,0 +1,815 @@
+
+    /**
+     * @alias oci_collection_element_get
+     * @return string|double|null
```
same

jensdenies commented:
@kocsismate Hi, i applied the requested changes and i think i've rebased master. Did i do it correct? I've ran the tests and not all of them pass, but i dont think it's related to my changes,.

kocsismate commented:
@jensdenies The fixes look good! However, I don't think you rebased to current master, because there should be a hash in the beginning of the arginfo file now. That possibly means you have to fetch current master from the `upstream` remote:

```
# Fetch and rebase the changes from the upstream
git pull upstream master --rebase
```
P.S. You don't need --rebase if you have configured git to use rebase by default

If you haven't done that before, you can set up the `upstream` remote like below:

`git remote add upstream git@github.php/php-src.git`

jensdenies commented:
Thanks @kocsismate, i think it worked. There's a hash in the file now.

kocsismate commented:
@jensdenies I've just merged your work! Congrats! As a follow-up, the `oci_internal_debug()` method should be removed. You can also try to fix the UNKNOWN defaults, if you're up to a new challenge. ^^

jensdenies commented:
Thanks @kocsismate for all your help! I noticed the oci_interna_debug function and emailed @cjbj about it. So i will make a new PR to remove that function, i’ll also make a PR to deprecate ocifetchinto, since i forgot it while deprecating the other aliases.
